File: server/src/game/Player.ts
// Player: hero, deck, hand, board, mana — các method nắm trong object (Rule of thumb Ownership).
import { CardNotInHandError, BoardFullError } from './errors.js';
import { MAX_HAND_SIZE, MAX_BOARD_SIZE, MAX_MANA } from './constants.js';
import { Deck } from './Deck.js';
import { Hero } from './Hero.js';
import { Minion } from './Minion.js';
import type { CardDefinition } from '@coincard/shared';

export class Player {
  private mana: number;
  private maxMana: number;
  private hand: CardDefinition[];
  private readonly board: Minion[] = [];

  constructor(
    public readonly playerId: string,
    public readonly name: string,
    public readonly hero: Hero,
    private readonly deck: Deck,
  ) {
    this.mana = 0;
    this.maxMana = 0;
    this.hand = [];
  }

  get id(): string {
    return this.playerId;
  }

  get currentMana(): number {
    return this.mana;
  }

  get currentMaxMana(): number {
    return this.maxMana;
  }

  get handCards(): Readonly<CardDefinition[]> {
    return this.hand.map((c) => ({ ...c }));
  }

  get handCount(): number {
    return this.hand.length;
  }

  get boardCards(): Readonly<Minion[]> {
    return this.board;
  }

  get boardCount(): number {
    return this.board.length;
  }

  get deckSize(): number {
    return this.deck.size;
  }

  get heroState(): Hero {
    return this.hero;
  }

  hasEnoughMana(manaCost: number): boolean {
    return this.mana >= manaCost;
  }

  spendMana(amount: number): void {
    this.mana -= amount;
  }

  refillMana(): void {
    this.mana = this.maxMana;
  }

  increaseMaxMana(): void {
    this.maxMana = Math.min(this.maxMana + 1, MAX_MANA);
  }

  drawCard(): CardDefinition {
    return this.deck.drawOne();
  }

  drawFromHand(count: number): CardDefinition[] {
    return this.deck.drawMultiple(count);
  }

  drawAndAddToHand(count: number): void {
    for (let i = 0; i < count; i++) {
      try {
        this.addToHand(this.deck.drawOne());
      } catch {
        // Fatigue — deck hết thì phải phải chịu damage theo rule (ở đây không cần draw)
      }
    }
  }

  // Opening-hand guarantee: đảm bảo tay đầu có ít nhất 1 lá rẻ để luôn có nước đi.
  // Đổi lá đắt nhất tay lấy lá rẻ nhất deck (số lượng tay không đổi).
  guaranteeCheapOpener(maxCost: number): void {
    const hasCheap = this.hand.some((c) => c.manaCost <= maxCost);
    if (hasCheap) {
      return;
    }
    let expensiveIndex = -1;
    for (let i = 0; i < this.hand.length; i++) {
      const card = this.hand[i];
      const current = expensiveIndex === -1 ? null : this.hand[expensiveIndex];
      if (!card) {
        continue;
      }
      if (!current || card.manaCost > current.manaCost) {
        expensiveIndex = i;
      }
    }
    if (expensiveIndex === -1) {
      return;
    }
    const removed = this.hand.splice(expensiveIndex, 1)[0];
    if (!removed) {
      return;
    }
    this.deck.returnToBottom(removed);
    const cheap = this.deck.drawCheapest(maxCost);
    if (cheap) {
      this.addToHand(cheap);
    } else {
      this.addToHand(removed);
    }
  }

  /** Dùng bởi GameEngine/Mulligan ban đầu. */
  getBoard(): Minion[] {
    return this.board;
  }

  addToHand(card: CardDefinition): void {
    if (this.hand.length >= MAX_HAND_SIZE) {
      return; // HS rule: deck full → burn (đồng bộ UI)
    }
    this.hand.push(card);
  }

  removeFromHand(instanceId: string): CardDefinition {
    const index = this.hand.findIndex((c) => c.id === instanceId);
    if (index === -1) {
      throw new CardNotInHandError();
    }
    const removed = this.hand.splice(index, 1)[0];
    if (!removed) {
      throw new CardNotInHandError();
    }
    return removed;
  }

  summonMinion(minion: Minion): void {
    if (this.board.length >= MAX_BOARD_SIZE) {
      throw new BoardFullError();
    }
    this.board.push(minion);
  }

  removeMinion(instanceId: string): Minion | null {
    const index = this.board.findIndex((m) => m.instanceId === instanceId);
    if (index === -1) {
      return null;
    }
    return this.board.splice(index, 1)[0] ?? null;
  }

  findMinion(instanceId: string): Minion | null {
    return this.board.find((m) => m.instanceId === instanceId) ?? null;
  }

  removeDeadMinions(): void {
    for (let i = this.board.length - 1; i >= 0; i--) {
      const minion = this.board[i];
      if (minion !== undefined && minion.isDead()) {
        this.board.splice(i, 1);
      }
    }
  }
}

File: server-java/src/main/java/vn/coincard/server/game/Player.java
package vn.coincard.server.game;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/** Player state and controlled mutation methods. */
public class Player {
  private int mana;
  private int maxMana;
  private int temporaryMana;
  private int fatigueDamage;
  private boolean heroPowerUsed;
  private final List<CardTypes.CardDefinition> hand = new ArrayList<>();
  private final List<Minion> board = new ArrayList<>();
  private int damageDealtValue;
  private int cardsPlayedValue;
  private int minionsSummonedValue;

  public Player(String playerId, String name, Hero hero, Deck deck) {
    this.playerId = playerId;
    this.name = name;
    this.hero = hero;
    this.deck = deck;
  }

  private final String playerId;
  private final String name;
  private final Hero hero;
  private final Deck deck;

  public String id() { return playerId; }
  public String name() { return name; }
  public int currentMana() { return mana + temporaryMana; }
  public int currentMaxMana() { return maxMana; }
  public int fatigueDamage() { return fatigueDamage; }
  public boolean heroPowerUsed() { return heroPowerUsed; }
  public List<CardTypes.CardDefinition> handCards() { return Collections.unmodifiableList(hand); }
  public int handCount() { return hand.size(); }
  public int boardCount() { return board.size(); }
  public int deckSize() { return deck.size(); }
  public Hero heroState() { return hero; }
  public int damageDealt() { return damageDealtValue; }
  public int cardsPlayed() { return cardsPlayedValue; }
  public int minionsSummoned() { return minionsSummonedValue; }

  public void recordDamage(int amount) {
    if (amount > 0) damageDealtValue += amount;
  }

  public void recordCardPlayed() { cardsPlayedValue++; }

  public void spendMana(int amount) {
    if (amount < 0 || amount > currentMana()) throw new IllegalArgumentException("Mana không hợp lệ.");
    int fromTemporary = Math.min(temporaryMana, amount);
    temporaryMana -= fromTemporary;
    mana -= amount - fromTemporary;
  }

  public void refillMana() {
    mana = maxMana;
    temporaryMana = 0;
    heroPowerUsed = false;
  }

  public void gainTemporaryMana(int amount) {
    if (amount < 0) throw new IllegalArgumentException("Temporary mana không hợp lệ.");
    temporaryMana = Math.min(Constants.MAX_MANA - mana, temporaryMana + amount);
  }

  public void markHeroPowerUsed() {
    if (heroPowerUsed) throw new GameException.HeroPowerAlreadyUsed();
    heroPowerUsed = true;
  }

  public void increaseMaxMana() { maxMana = Math.min(maxMana + 1, Constants.MAX_MANA); }

  public void drawAndAddToHand(int count) {
    for (int i = 0; i < count; i++) {
      if (deck.size() == 0) break;
      addToHand(deck.drawOne());
    }
  }

  /** Draws one card, burns it when the hand is full, or applies fatigue. */
  public void drawForTurn() {
    if (deck.size() == 0) {
      fatigueDamage++;
      hero.takeDamage(fatigueDamage);
      return;
    }
    addToHand(deck.drawOne());
  }

  /** Không expose mutable list trực tiếp — chỉ thay đổi qua behavior. */
  public List<Minion> getBoard() { return Collections.unmodifiableList(board); }

  public boolean addToHand(CardTypes.CardDefinition card) {
    if (hand.size() >= Constants.MAX_HAND_SIZE) return false; // burn
    hand.add(card);
    return true;
  }

  public CardTypes.CardDefinition findCardInHand(String instanceId) {
    return hand.stream().filter(c -> c.id().equals(instanceId)).findFirst().orElse(null);
  }

  public CardTypes.CardDefinition removeFromHand(String instanceId) {
    for (int i = 0; i < hand.size(); i++) {
      if (hand.get(i).id().equals(instanceId)) return hand.remove(i);
    }
    throw new GameException.CardNotInHand();
  }

  public void summonMinion(Minion minion) {
    if (!minion.getOwnerId().equals(id())) throw new IllegalArgumentException("Sai chủ sở hữu minion.");
    if (board.size() >= Constants.MAX_BOARD_SIZE) throw new GameException.BoardFull();
    board.add(minion);
    minionsSummonedValue++;
  }

  public Minion removeMinion(String instanceId) {
    for (int i = 0; i < board.size(); i++) {
      if (board.get(i).getInstanceId().equals(instanceId)) return board.remove(i);
    }
    return null;
  }

  public Minion findMinion(String instanceId) {
    return board.stream().filter(m -> m.getInstanceId().equals(instanceId)).findFirst().orElse(null);
  }

  public void removeDeadMinions() {
    board.removeIf(Minion::isDead);
  }

  /** Polymorph-style replace in place: no free slot needed, not counted as a summon. */
  public void replaceMinion(String instanceId, Minion replacement) {
    if (!replacement.getOwnerId().equals(id())) throw new IllegalArgumentException("Không thể thay minion.");
    for (int i = 0; i < board.size(); i++) {
      if (board.get(i).getInstanceId().equals(instanceId)) {
        board.set(i, replacement);
        return;
      }
    }
    throw new IllegalArgumentException("Không thể thay minion.");
  }

  record State(int mana, int maxMana, int temporaryMana, int fatigueDamage,
      boolean heroPowerUsed, List<CardTypes.CardDefinition> hand, List<Minion.State> board,
      Deck.State deck, Hero.State hero, int damageDealt, int cardsPlayed, int minionsSummoned) {
    State {
      hand = List.copyOf(hand);
      board = List.copyOf(board);
    }
  }

  State snapshotState() {
    return new State(mana, maxMana, temporaryMana, fatigueDamage, heroPowerUsed, hand,
        board.stream().map(Minion::snapshotState).toList(), deck.snapshotState(), hero.snapshotState(),
        damageDealtValue, cardsPlayedValue, minionsSummonedValue);
  }

  void restoreState(State state) {
    hand.clear();
    hand.addAll(state.hand());
    board.clear();
    for (Minion.State minionState : state.board()) {
      minionState.instance().restoreState(minionState);
      board.add(minionState.instance());
    }
    deck.restoreState(state.deck());
    hero.restoreState(state.hero());
    mana = state.mana();
    maxMana = state.maxMana();
    temporaryMana = state.temporaryMana();
    fatigueDamage = state.fatigueDamage();
    heroPowerUsed = state.heroPowerUsed();
    damageDealtValue = state.damageDealt();
    cardsPlayedValue = state.cardsPlayed();
    minionsSummonedValue = state.minionsSummoned();
  }
}
